nmigen_soc/periph.py

```diff
@@ -0,0 +1,59 @@
+from .memory import MemoryMap
+from . import event
+
+
+__all__ = ["PeripheralInfo"]
+
+
+class PeripheralInfo:
+    """Peripheral metadata.
+
+    A unified description of the local resources of a peripheral. It may be queried in order to
+    recover its memory windows, CSR registers and event sources.
+
+    Parameters
+    ----------
+    memory_map : :class:`MemoryMap`
+        Memory map of the peripheral.
+    irq : :class:`event.Source`
+        IRQ line of the peripheral. Optional.
+    """
+    def __init__(self, *, memory_map, irq=None):
+        if not isinstance(memory_map, MemoryMap):
+            raise TypeError("Memory map must be an instance of MemoryMap, not {!r}"
+                            .format(memory_map))
+        memory_map.freeze()
+        self._memory_map = memory_map
+
+        if irq is not None and not isinstance(irq, event.Source):
+            raise TypeError("IRQ line must be an instance of event.Source, not {!r}"
+                            .format(irq))
+        self._irq = irq
+
+    @property
+    def memory_map(self):
+        """Memory map.
+
+        Return value
+        ------------
+        A :class:`MemoryMap` describing the local address space of the peripheral.
+        """
+        return self._memory_map
+
+    @property
+    def irq(self):
+        """IRQ line.
+
+        Return value
+        ------------
+        An :class:`event.Source` used by the peripheral to request interrupts. If provided, its
+        event map describes local events.
+
+        Exceptions
+        ----------
+        Raises :exn:`NotImplementedError` if the peripheral info does not have an IRQ line.
+        """
+        if self._irq is None:
+            raise NotImplementedError("Peripheral info does not have an IRQ line"
```

Comment:
`NotImplementedError` is for omissions in code (think `TODO`s or `FIXME`s), not for cases where a part of an object is not present. I think there isn't anything that maps to this error particularly well, but I believe that `AttributeError` is not a bad choice here.

Reply:
Sounds good. In this case, getter errors in other parts of the codebase (e.g. [these](https://github.com/nmigen/nmigen-soc/search?q=NotImplementedError&unscoped_q=NotImplementedError)) should also be AttributeErrors for consistency.

Reply:
Hm, not all of those map cleanly to `AttributeError`. Let's merge this PR as-is and not block anything on that decision.